### PR TITLE
WIP: Add support for User authentication

### PR DESCRIPTION
**Pull request by gracechristina**
Will add the descriptions at the end of the day

**What this PR does / why we need it**:
This PR aims to implement user authentication for feast based on the RFC below:
https://docs.google.com/document/d/14-QBz9X8zK_aGAY2ti43a7PqMxvqp0035ec7QYgmEBM/edit#

**Does this PR introduce a user-facing change?**:
Yes, these are the changes caused by this Pull Request:

**Feast Core API**
New RPCs
AddMember
RemoveMember
ListMembers
AddRole
RemoveRole

**Feast Core Data Model**
**New table**
Users
Roles
UserProjectRole

**New classes**
Users
Roles

**Python SDK**
Addition of user auth configuration
New methods:
add_member()
remove_member()
list_members()
add_role()
remove_role()

### CODE REVIEW

File: core/src/main/java/feast/core/model/User.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+  @GeneratedValue(strategy = GenerationType.SEQUENCE)
+  @Column(name = "id", nullable = false, unique = true)
+  private Integer id;
+
+  // Name of the user
+  @Column(name = "name", nullable = false)
+  private String name;
+
+  //Password of the user
+//  @Column(name = "password_sha")
```
Comment (woop) [inline core/src/main/java/feast/core/model/User.java]: what is this about?

File: core/src/main/java/feast/core/model/User.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+  @ManyToMany(fetch = FetchType.EAGER)
+  @JoinTable(
+      name = "projects_members",
+      joinColumns = @JoinColumn(name = "user_id"),
+      inverseJoinColumns = @JoinColumn(name = "project_name")
+  )
+  Set<Project> projects;
+
+  public User() {
+    super();
```
Comment (woop) [inline core/src/main/java/feast/core/model/User.java]: In which case would we need this?

File: core/src/main/java/feast/core/model/User.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+  }
+
+  public User(String name) {
+    this.name = name;
+    this.projects = new HashSet<>();
+  }
+
+  public void setProject(Project project){
+    projects.add(project);
```
Comment (woop) [inline core/src/main/java/feast/core/model/User.java]: Shouldnt this be addProject?

File: core/src/main/java/feast/core/model/User.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+public class User {
+
+  // User Id
+  @Id
+  @GeneratedValue(strategy = GenerationType.SEQUENCE)
+  @Column(name = "id", nullable = false, unique = true)
+  private Integer id;
+
+  // Name of the user
+  @Column(name = "name", nullable = false)
```
Comment (woop) [inline core/src/main/java/feast/core/model/User.java]: Are we allowing the same name to be reused?

File: core/src/main/java/feast/core/service/AccessManagementService.java [hunk 5]
```diff
@@ -75,4 +85,48 @@ public void archiveProject(String name) {
...
+    return result;
+  }
+
+  @Transactional
+  public void addMember(String user_name, String project_name) {
+    Optional<Project> project = projectRepository.findById(project_name);
+    Project p = project.get();
+
+    if (userRepository.existsUserByName(user_name)) {
+      throw new IllegalArgumentException(String.format("User already exists: %s", user_name));
```
Comment (woop) [inline core/src/main/java/feast/core/service/AccessManagementService.java]: I wonder if this should be idempotent perhaps. Or at least not lead to an exception if the user has already been added. What do you think @gracechristina?

File: tests/e2e/basic-ingest-redis-serving.py [hunk 6]
```diff
@@ -128,22 +131,18 @@ def test_basic_retrieve_online_success(client, basic_dataframe):
                     }
                 )
             ],
-            feature_refs=[
-                "daily_transactions",
-                "total_transactions",
-            ],
+            feature_refs=["daily_transactions", "total_transactions",],
```
Comment (woop) [inline tests/e2e/basic-ingest-redis-serving.py]: Are you using Black?

File: core/src/main/java/feast/core/service/AccessManagementService.java [hunk 7]
```diff
@@ -75,4 +85,48 @@ public void archiveProject(String name) {
...
+    }
+    Project p = project.get();
+    Set<User> result = p.getProjectMembers();
+    return result;
+  }
+
+  @Transactional
+  public void addMember(String user_name, String project_name) {
+    Optional<Project> project = projectRepository.findById(project_name);
+    Project p = project.get();
```
Comment (woop) [inline core/src/main/java/feast/core/service/AccessManagementService.java]: Shouldn't only project members be able to grant access to projects to other members?

File: tests/e2e/basic-ingest-redis-serving.py [hunk 8]
```diff
@@ -23,31 +23,35 @@
...
-        "allow_dirty").lower() == "true" else False
+    return True if pytestconfig.getoption("allow_dirty").lower() == "true" else False
 
 
-@pytest.fixture(scope='module')
+@pytest.fixture(scope="module")
 def client(core_url, serving_url, allow_dirty):
     # Get client for core and serving
     client = Client(core_url=core_url, serving_url=serving_url)
     client.create_project(PROJECT_NAME)
```
Comment (woop) [inline tests/e2e/basic-ingest-redis-serving.py]: Where is the user setting their own identity (FEAST_USERNAME/FEAST_PASSWORD env vars)?

Comment (woop) [reply tests/e2e/basic-ingest-redis-serving.py]: We should probably seed the Core DB with a single admin user/pass in order to add more projects and members

File: core/src/main/java/feast/core/model/User.java (hunk 3 above)
Comment (gracechristina) [reply core/src/main/java/feast/core/model/User.java]: Renamed to addProject

File: core/src/main/java/feast/core/model/User.java (hunk 1 above)
Comment (gracechristina) [reply core/src/main/java/feast/core/model/User.java]: Uncommented

File: core/src/main/java/feast/core/model/User.java (hunk 2 above)
Comment (gracechristina) [reply core/src/main/java/feast/core/model/User.java]: It is not needed, removed

File: core/src/main/java/feast/core/model/User.java (hunk 4 above)
Comment (gracechristina) [reply core/src/main/java/feast/core/model/User.java]: Set the field to be unique

File: core/src/main/java/feast/core/service/AccessManagementService.java (hunk 5 above)
Comment (gracechristina) [reply core/src/main/java/feast/core/service/AccessManagementService.java]: Yes, agree. I think it is not needed anymore as the check of user name's uniqueness in the database has been done in the user model now.

File: core/src/main/java/feast/core/service/AccessManagementService.java (hunk 7 above)
Comment (gracechristina) [reply core/src/main/java/feast/core/service/AccessManagementService.java]: Added a check such that only the project member is allowed to grant access to other members

**Comment by gracechristina** [comment]
@woop I have added unit tests for the feast core side that is related to the user authentication

**Comment by ches** [comment]
Sorry to bring an RFC question here, I don't want to turn the PR into that level of discussion, but were we going to pull auth out from the Projects RFC into its own? @woop I think we mentioned privately a dedicated GH issue at least.

I'm triggered a bit already by Feast maintaining its own database of users and their credentials 😅

**Comment by woop** [comment]
> Sorry to bring an RFC question here, I don't want to turn the PR into that level of discussion, but were we going to pull auth out from the Projects RFC into its own? @woop I think we mentioned privately a dedicated GH issue at least.
> 
> I'm triggered a bit already by Feast maintaining its own database of users and their credentials

Happy to do that. There is definitely a drift that occurs from RFC to actual implementation, so we should also decide what eventually becomes the source of truth. Perhaps a flow like
1. RFC
2. PR (with active discussion, and documentation)
Then Gitbook becomes the source of truth and the RFC can go out of date.

In this case since auth & projects were mixed in together, I think creating an issue to externalize it makes sense. 

In terms of managing users & credentials. It's definitely not ideal, and we'd want to offload this to an external service ideally. I think the intent here was to have a basic form off auth out to protect our existing users while we implement something that is more integrated.

**Comment by woop** [comment]
@ches Added the issue here: https://github.com/gojek/feast/issues/504

**Comment by ches** [comment]
Completely understand and agree about the flow of the source of truth, in general.

**Comment by woop** [comment]
Closing this PR until we settle #504. We can submit a new PR at that point.

**Pull request closed without merging**